zebra-consensus/src/block/subsidy/funding_streams.rs

```diff
@@ -50,3 +48,116 @@ pub fn funding_stream_values(
+}
+
+/// Returns the position in the address slice for each funding stream
+/// as described in [protocol specification §7.10][7.10]
+///
+/// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
+fn funding_stream_address_index(height: Height, network: Network) -> usize {
+    let num_addresses = match network {
+        Network::Mainnet => FUNDING_STREAMS_NUM_ADDRESSES_MAINNET,
+        Network::Testnet => FUNDING_STREAMS_NUM_ADDRESSES_TESTNET,
+    };
+
+    let index = 1u32
+        .checked_add(funding_stream_address_period(height, network))
+        .expect("no overflow should happen in this sum")
+        .checked_sub(funding_stream_address_period(
+            FUNDING_STREAM_HEIGHT_RANGES.get(&network).unwrap().start,
+            network,
+        ))
+        .expect("no overflow should happen in this sub") as usize;
+
+    assert!(index > 0 && index <= num_addresses);
+    // spec formula will output an index starting at 1 but
+    // Zebra indices for addresses start at zero, return converted.
+    index - 1
+}
+
+/// Return the address corresponding to given height, network and funding stream receiver.
+pub fn funding_stream_address(
+    height: Height,
+    network: Network,
+    receiver: FundingStreamReceiver,
+) -> Address {
+    let index = funding_stream_address_index(height, network);
+    let address = &FUNDING_STREAM_ADDRESSES
+        .get(&network)
+        .expect("there is always another hash map as value for a given valid network")
+        .get(&receiver)
+        .expect("in the inner hash map there is always a vector of strings with addresses")[index];
+    Address::from_str(address).expect("Address should deserialize")
+}
+
+/// Given a founders reward address, create a script and check if it is the same
+/// as the given lock_script as described in [protocol specification §7.10][7.10]
+///
+/// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams.
+pub fn check_script_form(lock_script: Script, address: Address) -> bool {
+    let mut address_hash = address
+        .zcash_serialize_to_vec()
+        .expect("we should get address bytes here");
+
+    address_hash = address_hash[2..22].to_vec();
+    address_hash.insert(0, OpCode::Push20Bytes as u8);
+    address_hash.insert(0, OpCode::Hash160 as u8);
+    address_hash.insert(address_hash.len(), OpCode::Equal as u8);
+    if lock_script.as_raw_bytes().len() == address_hash.len()
+        && lock_script == Script::new(&address_hash)
+    {
+        return true;
+    }
+    false
+}
+
+/// Returns a list of outputs in `Transaction`, which have a script address equal to `Address`
+/// and a value equal to `Amount`.
+pub fn find_output_with_address_and_amount(
+    transaction: &Transaction,
+    address: Address,
+    amount: Amount<NonNegative>,
+) -> Vec<Output> {
+    transaction
+        .outputs()
+        .iter()
+        .filter(|o| check_script_form(o.lock_script.clone(), address) && o.value == amount)
```

Comment:
Minor optimization: we could remove the `clone` by changing `check_script_form` to receive a `&Script`.

```suggestion
        .filter(|o| check_script_form(o.lock_script, address) && o.value == amount)
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/3040/commits/d0e26e93c143b1163dbef32cd025e5b6cae2dacc